sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,341 @@
+from ._target_encoder_fast import _fit_encoding_fast
+from ._target_encoder_fast import _fit_encoding_fast_auto_smooth
+from ..utils.validation import _check_y, check_consistent_length
+from ..utils.multiclass import type_of_target
+from ..utils._param_validation import Interval, StrOptions
+
+
+class TargetEncoder(OneToOneFeatureMixin, _BaseEncoder):
+    """Target Encoder for regression and classification targets.
+
+    Each category is encoded based on a shrinked estimate of the average target
+    values for observations belonging to the category. The encoding scheme mixes
+    the global target mean with the target mean conditioned on the value of the
+    category. [MIC]_
+
+    :class:`TargetEncoder` considers missing values, such as `np.nan` or `None`,
+    as another category and encodes them like any other category. Categories
+    that are not seen during :meth:`fit` are encoded with the target mean, i.e.
+    `target_mean_`.
+
+    Read more in the :ref:`User Guide <target_encoder>`.
+
+    .. note::
+        `fit(X, y).transform(X)` does not equal `fit_transform(X, y)` because a
+        cross-validation scheme is used in `fit_transform` for encoding. See the
+        :ref:`User Guide <target_encoder>`. for details.
+
+    .. versionadded:: 1.3
+
+    Parameters
+    ----------
+    categories : "auto" or a list of array-like, default="auto"
+        Categories (unique values) per feature:
+
+        - `"auto"` : Determine categories automatically from the training data.
+        - list : `categories[i]` holds the categories expected in the i-th column. The
+          passed categories should not mix strings and numeric values within a single
+          feature, and should be sorted in case of numeric values.
+
+        The used categories is stored in the `categories_` fitted attribute.
+
+    target_type : {"auto", "continuous", "binary"}, default="auto"
+        Type of target.
+
+        - `"auto"` : Type of target is inferred with
+          :func:`~sklearn.utils.multiclass.type_of_target`.
+        - `"continuous"` : Continuous target
+        - `"binary"` : Binary target
+
+        .. note::
+            The type of target inferred with `"auto"` may not be the desired target
+            type used for modeling. For example, if the target consistent of integers
+            between 0 and 100, then :func:`~sklearn.utils.multiclass.type_of_target`
+            will infer the target as `"multiclass"`. In this case, setting
+            `target_type="continuous"` will understand the target as a regression
+            problem. The `target_type_` attribute gives the target type used by the
+            encoder.
+
+    smooth : "auto" or float, default="auto"
+        The amount of mixing of the categorical encoding with the global target mean. A
+        larger `smooth` value will put more weight on the global target mean.
+        If `"auto"`, then `smooth` is set to an empirical Bayes estimate.
```

Comment:
I think we should allow for an `n_features` length parametrization here (e.g. list and/or dict with feature names as keys): the optimal smoothing for one feature is not necessarily optimal for others.

This can be done as follow-up PR though. It feels border-line YAGNI to me. In the mean time it possible to use a column-transformer to configure a per-feature `TargetEncoder` instance with dedicated `smooth` values per features.